File: python3/Stacks/implement_a_queue_using_a_stack.py
class Queue:
    def __init__(self):
        self.enqueue_stack = []
        self.dequeue_stack = []

    def enqueue(self, x: int) -> None:
        self.enqueue_stack.append(x)

    def transfer_enqueue_to_dequeue(self) -> None:
        # If the dequeue stack is empty, push all elements from the enqueue stack
        # onto the dequeue stack. This ensures the top of the dequeue stack
        # contains the most recent value.
        if not self.dequeue_stack:
            while self.enqueue_stack:
                self.dequeue_stack.append(self.enqueue_stack.pop())

    def dequeue(self) -> int:
        self.transfer_enqueue_to_dequeue()
        # Pop and return the value at the top of the dequeue stack.
        return self.dequeue_stack.pop() if self.dequeue_stack else None

    def peek(self) -> int:
        self.transfer_enqueue_to_dequeue()
        return self.dequeue_stack[-1] if self.dequeue_stack else None

File: kotlin/Stacks/ImplementAQueueUsingAStack.kt
import java.util.Stack

class Queue {

    private val enqueueStack = Stack<Int>()
    private val dequeueStack = Stack<Int>()

    fun enqueue(x: Int) {
        enqueueStack.push(x)
    }

    private fun transferEnqueueToDequeue() {
        // If the dequeue stack is empty, push all elements from the enqueue stack
        // onto the dequeue stack. This ensures the top of the dequeue stack
        // contains the most recent value.
        if (dequeueStack.isEmpty()) {
            while (enqueueStack.isNotEmpty()) {
                dequeueStack.push(enqueueStack.pop())
            }
        }
    }

    fun dequeue(): Int {
        transferEnqueueToDequeue()
        // Pop and return the value at the top of the dequeue stack.
        return if (dequeueStack.isNotEmpty()) {
            dequeueStack.pop()
        } else {
            -1
        }
    }

    fun peek(): Int {
        transferEnqueueToDequeue()
        return if (dequeueStack.isNotEmpty()) {
            dequeueStack.peek()
        } else {
            -1
        }
    }
}
